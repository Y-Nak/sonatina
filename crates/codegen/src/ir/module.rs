use cranelift_entity::{entity_impl, PrimaryMap, SecondaryMap};

use crate::ir::Function;

use crate::isa::TargetIsa;

#[derive(Debug)]
pub struct Module {
    /// Target ISA of the module.
    pub isa: TargetIsa,

    /// Holds all function declared in the contract.
    pub funcs: PrimaryMap<FuncRef, Function>,

    pub func_attributes: SecondaryMap<FuncRef, FuncAttribute>,
}

impl Module {}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct FuncAttribute {
    pub linkage: Linkage,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct FuncRef(u32);
entity_impl!(FuncRef);

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
/// Linkage of the function.
pub enum Linkage {
    /// The function is defined in the contract, and can be called from another accounts.
    Public,

    /// The function is defined in the contract, and can NOT be called from another accounts.
    Private,

    /// The function is defined outside of the contract.
    External,
}
