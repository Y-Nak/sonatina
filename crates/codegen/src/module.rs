use cranelift_entity::{entity_impl, PrimaryMap, SecondaryMap};

use crate::{Function, Signature};

#[derive(Debug, Clone)]
pub struct Module {
    /// Hold all function declared in the contract.
    pub funcs: PrimaryMap<FuncRef, Function>,

    pub func_attributes: SecondaryMap<FuncRef, FuncAttribute>,
}

impl Module {
    pub fn declare_function(&mut self, name: String, sig: Signature, linkage: Linkage) -> FuncRef {
        let func = Function::new(name, sig);
        let func_ref = self.funcs.push(func);
        self.func_attributes[func_ref] = FuncAttribute { linkage };
        func_ref
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct FuncAttribute {
    pub linkage: Linkage,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct FuncRef(u32);
entity_impl!(FuncRef);

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
/// Linkage of the function.
pub enum Linkage {
    /// The function is defined in the contract, and can be called from another accounts.
    Public,

    /// The function is defined in the contract, and can NOT be called from another accounts.
    Private,

    /// The function is defined outside of the contract.
    External,
}
