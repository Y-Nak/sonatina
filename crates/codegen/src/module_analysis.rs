use cranelift_entity::{entity_impl, PrimaryMap, SecondaryMap};
use dashmap::DashMap;
use rustc_hash::FxHashSet;
use sonatina_ir::{module::FuncRef, Module};

#[derive(Debug, Clone, Default)]
pub struct CallGraph {
    nodes: SecondaryMap<FuncRef, Node>,
}

impl CallGraph {
    /// Build a call graph from a module.
    pub fn build_graph(module: &Module) -> Self {
        let d_nodes = DashMap::new();
        module.func_store.par_for_each(|func_ref, func| {
            let mut callees = FxHashSet::default();
            for block in func.layout.iter_block() {
                for inst_id in func.layout.iter_inst(block) {
                    if let Some(call) = func.dfg.call_info(inst_id) {
                        callees.insert(call.callee());
                    }
                }
            }

            let mut callees: Vec<_> = callees.into_iter().collect();
            callees.sort_unstable();
            let node = Node { callees };

            d_nodes.insert(func_ref, node);
        });

        let mut nodes = SecondaryMap::new();
        for (func_ref, node) in d_nodes {
            nodes[func_ref] = node;
        }
        CallGraph { nodes }
    }

    /// Get the callees of a function.
    pub fn callee_of(&self, func_ref: FuncRef) -> &[FuncRef] {
        &self.nodes[func_ref].callees
    }
}

#[derive(Debug, Clone, Default)]
struct Node {
    callees: Vec<FuncRef>,
}

/// Represents the strongly connected components of a call graph in a module.
pub struct CallGraphSccs {
    scc_map: SecondaryMap<FuncRef, SccRef>,
    scc_store: PrimaryMap<SccRef, SccInfo>,
}

impl CallGraphSccs {
    pub fn scc_of(&self, func_ref: FuncRef) -> &SccInfo {
        let scc_ref = self.scc_map[func_ref];
        &self.scc_store[scc_ref]
    }
}

#[derive(Debug, Clone)]
pub struct SccInfo {
    /// Whether the SCC is a true cycle.
    /// NOTE: The SCC fomes a true cycle if it contains more than one function
    /// or a function that calls itself.
    pub is_cycle: bool,

    /// The functions in the SCC.
    pub scc: FxHashSet<FuncRef>,
}

#[derive(Debug)]
pub struct SccBuilder {
    scc_map: SecondaryMap<FuncRef, SccRef>,
    scc_store: PrimaryMap<SccRef, SccInfo>,
    stack: Vec<FuncRef>,
    nodes: SecondaryMap<FuncRef, NodeState>,
    next_index: usize,
}

impl SccBuilder {
    pub fn compute_scc(mut self, call_graph: &CallGraph) -> CallGraphSccs {
        for func_ref in call_graph.nodes.keys() {
            if !self.nodes[func_ref].visited {
                self.strong_component(func_ref, call_graph);
            }
        }

        CallGraphSccs {
            scc_map: self.scc_map,
            scc_store: self.scc_store,
        }
    }

    fn strong_component(&mut self, func_ref: FuncRef, call_graph: &CallGraph) {
        let index = self.next_index;
        self.next_index += 1;
        let node_info = NodeState {
            index,
            lowlink: index,
            on_stack: true,
            visited: true,
        };
        self.nodes[func_ref] = node_info;
        self.stack.push(func_ref);

        let mut is_trivial_cycle = false;

        for &callee in call_graph.callee_of(func_ref) {
            is_trivial_cycle |= callee == func_ref;

            if !self.nodes[callee].visited {
                self.strong_component(callee, call_graph);
                self.nodes[func_ref].lowlink =
                    self.nodes[func_ref].lowlink.min(self.nodes[callee].lowlink);
            } else if self.nodes[callee].on_stack {
                self.nodes[func_ref].lowlink =
                    self.nodes[func_ref].lowlink.min(self.nodes[callee].index);
            }
        }

        if self.nodes[func_ref].index == self.nodes[func_ref].lowlink {
            let mut scc = FxHashSet::default();

            loop {
                let top = self.stack.pop().unwrap();
                self.nodes[top].on_stack = false;
                scc.insert(top);

                if top == func_ref {
                    break;
                }
            }

            let scc_ref = self.scc_store.push(SccInfo {
                scc: FxHashSet::default(),
                is_cycle: scc.len() > 1 || is_trivial_cycle,
            });

            for &func_ref in &scc {
                self.scc_map[func_ref] = scc_ref;
            }

            self.scc_store[scc_ref].scc = scc;
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
struct SccRef(u32);
entity_impl!(SccRef);

#[derive(Default, Debug, Clone)]
struct NodeState {
    index: usize,
    lowlink: usize,
    on_stack: bool,
    visited: bool,
}
