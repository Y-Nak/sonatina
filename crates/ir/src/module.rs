use std::sync::{Arc, Mutex};

use cranelift_entity::{entity_impl, PrimaryMap};

use crate::Function;

use crate::{global_variable::GlobalVariableStore, isa::TargetIsa, types::TypeStore};

use super::Linkage;

#[derive(Debug)]
pub struct Module {
    /// Holds all function declared in the contract.
    pub funcs: PrimaryMap<FuncRef, Function>,

    pub ctx: ModuleCtx,
}

impl Module {
    #[doc(hidden)]
    pub fn new(isa: TargetIsa) -> Self {
        Self {
            funcs: PrimaryMap::default(),
            ctx: ModuleCtx::new(isa),
        }
    }

    /// Returns `func_ref` in the module.
    pub fn iter_functions(&self) -> impl Iterator<Item = FuncRef> {
        self.funcs.keys()
    }

    /// Returns `true` if the function has external linkage.
    pub fn is_external(&self, func_ref: FuncRef) -> bool {
        self.funcs[func_ref].sig.linkage() == Linkage::External
    }
}

#[derive(Debug, Clone)]
pub struct ModuleCtx {
    pub isa: TargetIsa,
    // TODO: Consider using `RwLock` instead of `Mutex`.
    type_store: Arc<Mutex<TypeStore>>,
    // TODO: Consider using `RwLock` instead of `Mutex`.
    gv_store: Arc<Mutex<GlobalVariableStore>>,
}

impl ModuleCtx {
    pub fn new(isa: TargetIsa) -> Self {
        Self {
            isa,
            type_store: Arc::new(Mutex::new(TypeStore::default())),
            gv_store: Arc::new(Mutex::new(GlobalVariableStore::default())),
        }
    }

    pub fn with_ty_store<F, R>(&self, f: F) -> R
    where
        F: FnOnce(&TypeStore) -> R,
    {
        f(&self.type_store.lock().unwrap())
    }

    pub fn with_ty_store_mut<F, R>(&self, f: F) -> R
    where
        F: FnOnce(&mut TypeStore) -> R,
    {
        f(&mut self.type_store.lock().unwrap())
    }

    pub fn with_gv_store<F, R>(&self, f: F) -> R
    where
        F: FnOnce(&GlobalVariableStore) -> R,
    {
        f(&self.gv_store.lock().unwrap())
    }

    pub fn with_gv_store_mut<F, R>(&self, f: F) -> R
    where
        F: FnOnce(&mut GlobalVariableStore) -> R,
    {
        f(&mut self.gv_store.lock().unwrap())
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct FuncRef(u32);
entity_impl!(FuncRef);
