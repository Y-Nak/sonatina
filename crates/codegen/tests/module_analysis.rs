mod common;

use std::fmt::Write;

use dir_test::{dir_test, Fixture};
use indexmap::IndexSet;
use sonatina_codegen::module_analysis::{analyze_module, ModuleInfo, SccRef};
use sonatina_ir::module::ModuleCtx;
use sonatina_parser::ParsedModule;

#[dir_test(
    dir: "$CARGO_MANIFEST_DIR/test_files/module_analysis/",
    glob: "*.sntn"
    loader: common::parse_module,
)]
fn test(fixture: Fixture<ParsedModule>) {
    let module = &fixture.content().module;
    let module_info = analyze_module(module);
    let info_string = dump_module_info(&module.ctx, &module_info);

    snap_test!(info_string, fixture.path());
}

fn dump_module_info(ctx: &ModuleCtx, info: &ModuleInfo) -> String {
    let mut s = String::new();
    // Write access pattern.
    writeln!(&mut s, "ModuleDependency: {}\n", info.access_pattern).unwrap();

    let mut sccs: IndexSet<SccRef> = IndexSet::default();

    // Write SCCs.
    for func_ref in info.call_graph.funcs() {
        let scc_ref = info.scc.scc_ref(func_ref);
        if sccs.insert(info.scc.scc_ref(func_ref)) {
            write!(&mut s, "SCC: [").unwrap();
            let mut components: Vec<_> = info
                .scc
                .scc_info(scc_ref)
                .components
                .iter()
                .copied()
                .collect();
            components.sort_unstable();
            let mut components = components.into_iter();

            if let Some(func_ref) = components.next() {
                ctx.func_sig(func_ref, |sig| {
                    write!(&mut s, "`%{}`", sig.name()).unwrap();
                });
            }

            for func_ref in components {
                ctx.func_sig(func_ref, |sig| {
                    write!(&mut s, ", `%{}`", sig.name()).unwrap();
                });
            }

            writeln!(&mut s, "]").unwrap();
        }
    }
    writeln!(&mut s).unwrap();

    for func_ref in info.call_graph.funcs() {
        let func_info = info.func_info.get(&func_ref).unwrap();
        ctx.func_sig(func_ref, |sig| {
            writeln!(&mut s, "`%{}` = {:?}", sig.name(), func_info).unwrap();
        });
    }

    s
}
