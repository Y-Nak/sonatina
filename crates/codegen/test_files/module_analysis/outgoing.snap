---
source: crates/codegen/tests/module_analysis.rs
input_file: test_files/module_analysis/outgoing.sntn
---
ModuleDependency: OutgoingOnly

SCC: [`%fx`]
SCC: [`%f1`, `%f2`]
SCC: [`%f3`]
SCC: [`%f4`]
SCC: [`%f5`]
SCC: [`%f6`]

`%fx` = FuncInfo { is_recursive: true, flow: Bidirectional, is_leaf: false }
`%f1` = FuncInfo { is_recursive: true, flow: OutgoingOnly, is_leaf: false }
`%f2` = FuncInfo { is_recursive: true, flow: OutgoingOnly, is_leaf: false }
`%f3` = FuncInfo { is_recursive: true, flow: Closed, is_leaf: false }
`%f4` = FuncInfo { is_recursive: false, flow: OutgoingOnly, is_leaf: false }
`%f5` = FuncInfo { is_recursive: false, flow: OutgoingOnly, is_leaf: false }
`%f6` = FuncInfo { is_recursive: false, flow: OutgoingOnly, is_leaf: false }
